common/data_refinery_common/rna_seq.py

```diff
@@ -78,40 +77,37 @@ def get_quant_results_for_experiment(experiment: Experiment, filter_old_versions
     """Returns a queryset of salmon quant results from `experiment`."""
     # Subquery to calculate quant results
     # https://docs.djangoproject.com/en/2.2/ref/models/expressions/#subquery-expressions
-
-    # Salmon version gets saved as what salmon outputs, which includes this prefix.
-    current_salmon_version = "salmon " + get_env_variable("SALMON_VERSION", "0.13.1")
+    all_results = ComputationalResult.objects.filter(sample__in=experiment.samples.all())
 
     if filter_old_versions:
-        eligible_results = ComputationalResult.objects.prefetch_related("organism_index").filter(
-            organism_index__salmon_version=current_salmon_version
+        # Salmon version gets saved as what salmon outputs, which includes this prefix.
+        current_salmon_version = "salmon " + get_env_variable("SALMON_VERSION", "0.13.1")
+        organisms = experiment.organisms.all()
+        organism_indices = OrganismIndex.objects.filter(
+            salmon_version=current_salmon_version, organism__in=organisms
         )
-    else:
-        eligible_results = ComputationalResult.objects.all()
+        all_results.filter(organism_index__id__in=organism_indices.values("id"))
 
-    # A result is only eligible to be used if it actually got uploaded.
-    eligible_results = eligible_results.select_related("computedfile").filter(
-        computedfile__s3_bucket__isnull=False, computedfile__s3_key__isnull=False
-    )
-
-    # Calculate the computational results sorted that are associated with a given sample (
-    # referenced from the top query)
-    newest_computational_results = eligible_results.filter(
-        samples=OuterRef("id"), processor__name=ProcessorEnum.SALMON_QUANT.value["name"],
-    ).order_by("-created_at")
-
-    # Annotate each sample in the experiment with the id of the most recent computational result
-    computational_results_ids = (
-        experiment.samples.all()
-        .annotate(
-            latest_computational_result_id=Subquery(newest_computational_results.values("id")[:1])
+        all_results.select_related("computedfile").filter(
+            computedfile__s3_bucket__isnull=False, computedfile__s3_key__isnull=False
         )
-        .filter(latest_computational_result_id__isnull=False)
-        .values_list("latest_computational_result_id", flat=True)
-    )
 
-    # return the computational results that match those ids
-    return ComputationalResult.objects.all().filter(id__in=computational_results_ids)
+    def get_sample_id_set(result):
+        return {sample.id for sample in result.samples.all()}
+
+    latest_results = set()
+    for k, group in groupby(all_results, get_sample_id_set):
```

Comment:
Do you have to sort before you `groupby` here? The `itertools` docs say that `groupby` doesn't automatically sort beforehand: https://docs.python.org/3/library/itertools.html#itertools.groupby

> The operation of groupby() is similar to the uniq filter in Unix. It generates a break or new group every time the value of the key function changes (which is why it is usually necessary to have sorted the data using the same key function).

Reply:
Eww WTF? Why does python's groupby suck so bad? I miss clojure

Good catch though, I didn't read the docs carefully enough and made assumptions about how easy it should be to use.